grid-compose/internal/app.go

```diff
@@ -0,0 +1,372 @@
+			return err
+		}
+
+		return nil
+	}
+
+	// print for better formatting
+	fmt.Printf("\n%s\n", output.String())
+	return nil
+}
+
+func parsePsFlags(flags *pflag.FlagSet) (bool, string, error) {
+	verbose, err := flags.GetBool("verbose")
+	if err != nil {
+		return verbose, "", err
+	}
+
+	outputFile, err := flags.GetString("output")
+	if err != nil {
+		return verbose, outputFile, err
+	}
+
+	return verbose, outputFile, nil
+}
+
+// Down cancels all the deployments
+func (a *App) Down() error {
+	for networkName := range a.Config.DeploymentData {
+		projectName := a.getProjectName(networkName)
+		log.Info().Str("projectName", projectName).Msg("canceling deployments")
+		if err := a.Client.CancelByProjectName(projectName); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *App) generateNetworks() map[string]*workloads.ZNet {
+	zNets := make(map[string]*workloads.ZNet, 0)
+	defNetName := utils.GenerateDefaultNetworkName(a.Config.Services)
+	for networkName := range a.Config.DeploymentData {
+		if networkName == defNetName {
+			zNets[networkName] = &workloads.ZNet{
+				Name: networkName,
+				IPRange: gridtypes.NewIPNet(net.IPNet{
+					IP:   net.IPv4(10, 20, 0, 0),
+					Mask: net.CIDRMask(16, 32),
+				}),
+				AddWGAccess: false,
+			}
+		} else {
+			network := a.Config.Networks[networkName]
+			zNets[networkName] = &workloads.ZNet{
+				Name:         network.Name,
+				Description:  network.Description,
+				IPRange:      gridtypes.NewIPNet(utils.GenerateIPNet(network.IPRange.IP, network.IPRange.Mask)),
+				AddWGAccess:  network.AddWGAccess,
+				MyceliumKeys: network.MyceliumKeys,
+			}
+		}
+	}
+
+	return zNets
+}
+
+func (a *App) getProjectName(key string) string {
+	return fmt.Sprintf("compose/%v/%v", a.Client.TwinID, key)
+}
+
+func (a *App) getDeploymentName(key string) string {
+	return fmt.Sprintf("dl_%v", key)
+}
+
+func (a *App) loadCurrentNodeDeployments(projectName string) error {
+	contracts, err := a.Client.ContractsGetter.ListContractsOfProjectName(projectName, true)
+	if err != nil {
+		return err
+	}
+
+	var nodeID uint32
+
+	for _, contract := range contracts.NodeContracts {
+		contractID, err := strconv.ParseUint(contract.ContractID, 10, 64)
+		if err != nil {
+			return err
+		}
+
+		nodeID = contract.NodeID
+		a.checkIfExistAndAppend(nodeID, contractID)
+	}
+
+	return nil
+}
+
+func (a *App) checkIfExistAndAppend(node uint32, contractID uint64) {
+	for _, n := range a.Client.State.CurrentNodeDeployments[node] {
+		if n == contractID {
+			return
+		}
+	}
+
+	a.Client.State.CurrentNodeDeployments[node] = append(a.Client.State.CurrentNodeDeployments[node], contractID)
+}
+
+// TODO: Calculate total MRU and SRU while populating the deployment data
+func (a *App) generateMissingNodes(ctx context.Context) error {
+	for _, deploymentData := range a.Config.DeploymentData {
+		if deploymentData.NodeID != 0 {
+			continue
+		}
+
+		// freeCRU is not in NodeFilter?
+		var freeMRU, freeSRU uint64
```

Comment:
yes freeCRU is not a filter. due to cores overprovisioning mulitple vms can use the same cpu cores, use totalCRU instead just making sure the node has the cores you need